src/machine/machine_avr.go

```diff
@@ -58,3 +58,68 @@ func (p GPIO) Get() bool {
 		return (val > 0)
 	}
 }
+
+// InitPWM initializes the registers needed for PWM.
+func InitPWM() {
+	// use waveform generation
+	*avr.TCCR0A |= avr.TCCR0A_WGM00
+
+	// set timer 0 prescale factor to 64
+	*avr.TCCR0B |= avr.TCCR0B_CS01 | avr.TCCR0B_CS00
+
+	// set timer 1 prescale factor to 64
+	*avr.TCCR1B |= avr.TCCR1B_CS11
+
+	// put timer 1 in 8-bit phase correct pwm mode
+	*avr.TCCR1A |= avr.TCCR1A_WGM10
+
+	// set timer 2 prescale factor to 64
+	*avr.TCCR2B |= avr.TCCR2B_CS22
+
+	// configure timer 2 for phase correct pwm (8-bit)
+	*avr.TCCR2A |= avr.TCCR2A_WGM20
+}
+
+// Configure configures a PWM pin for output.
+func (pwm PWM) Configure() {
+	if pwm.Pin < 8 {
+		*avr.DDRD |= 1 << pwm.Pin
+	} else {
+		*avr.DDRB |= 1 << (pwm.Pin - 8)
+	}
+}
+
+// Set turns on the duty cycle for a PWM pin using the provided value. On the AVR this is normally a
+// 8-bit value ranging from 0 to 255.
+func (pwm PWM) Set(value uint16) {
```

Comment:
I see you have switched to 16-bit numbers for the duty cycle, but that wasn't entirely as I intended.
See [here](https://github.com/aykevl/tinygo/wiki/Machine-API) for details on my proposal (comments welcome). To be concrete, my idea for this PR was to use `uint8` everywhere as it's the fastest on AVR, but only convert to the full 16-bit range across the `PWM.Set()` boundary (using `.Set(uint16(value)*256)`). This way the PWM implementation can support full 16-bit PWM when available and fall back to 8-bit PWM (after dividing by 256) on hardware that doesn't have this precision. Doing it this way makes sure the caller can work without having to know how many bits the PWM has.

This means that on the AVR, the 8 bits PWM would be set using e.g. `OCR2B = avr.RegValue(value >> 8)` and 16-bits PWM can use something like `OCR1AL = avr.RegValue(value); OCR1AH = avr.RegValue(value >> 8);` for transparent 16-bit operation.

Sorry to be nitpicking here again, but I would really like to get the design right first (I missed it yesterday). If you have a different idea of how this should be implemented, please let me know.

Reply:
Took me a moment to follow what you were saying here, but then I noticed the wiki page. Did not realize that was there.

The only question I have is, is it better to have a single 16-bit scale that Arduino programmers have to write code that is different then the 0-254 they expect? My intent was to try to use the scaling that fits the docs for the board you are working with.

That is the reason I implemented it in this way, assuming that many people will simply enter the 8-bit value when "porting" Arduino code to TinyGo. We can enter a GoDoc comment to inform people they should use the 16-bit scaled value to try to steer them towards the correct value for their specific hardware.

Reply:
Yeah I wasn't sure how to express it clearly but glad you got it.

"Arduino programmers expect it" is not a very convincing argument to me, but googling PWM a bit more I see a different issue (obvious in hindsight): most MCUs are physically not able to reach a higher frequency without sacrifices on the frequency (because the clock speed is usually around 16MHz or not much higher). So maybe it's best to simply use 8 bits values for now and add `Set16()` and `SetFrequency()` (or something like that) later on for programs that require more precision.

In other words: I think simply using `uint8` is best after all. I can change that for you when merging (it's a fairly trivial change) or you can add a new commit. Both are fine to me.

Reply:
Well, I think the nrf ARM devices use 16-bit values, no? In which case we will have this same question immediately on the next platform. Seems like we should do the uint16 value as you first proposed.

Reply:
I've looked into it and it seems that it uses 16-bit values. But the PWM there is very different from the one on the AVR (it even has DMA!). For example, it allows to set a custom 'top' value, meaning the max value that is supported (the dutycycle is then `duty/top*100%`). Additionally, there is a divisor based on the clock frequency (where the clock frequency is fixed to 16MHz), so you can get 16MHz, 8MHz, 4MHz, etc. Oh, and a single PWM module controls multiple pins, so it isn't possible to set the top value different for each pin.

Still, the issue with frequency vs accuracy remains, and is fairly fundamental (limited by clock speed). A possible implementation for nrf would be fixing the divisor to 16MHz, the top to 255, and control the PWM with the duty cycle (called a channel on the nrf) to get an equivalent implementation as for the Arduino. Programs that need more flexibility can use the registers directly or use a different package.

In other words, I think using `uint8` is fine for nrf.

Reply:
In that case, do you mind adding another commit to the PR to make it how you think it should be? I can test on my LED setup before merge if that is helpful, otherwise just land it!

Reply:
I have made two implementations, one using `uint16` as I originally intended and one using `uint8` like you wrote originally: https://github.com/aykevl/tinygo/commit/56d843515fd726881115dcf6049bfd64155a5061 and https://github.com/aykevl/tinygo/commit/f2244539d4fd9e4e0e3c2bf069c9e6cfea1020a7.

They compile down to the exact same code, so when used like this it's a zero-cost abstraction to use 16-bit values. However, using `uint16` in the color calculations increases code size a bit as the calculations are done in 16 bits instead of 8.
Both implementations are fine by me, by now I'm not entirely sure what the best API is...
Do you have a preference?

> I can test on my LED setup before merge if that is helpful, otherwise just land it!

If you want, that would be useful (just to be sure).

Reply:
I vote for 16 bit duty cycle, since that will let us stick with that for pretty much all platforms in the immediate future.

Reply:
After sleeping on it, I think I still prefer the `unint16` approach you defined in the wiki doc.